[Liquid Glass] [macOS] Tab snapshots overlap with tab bar when swiping back/forwards

whsieh opened the pull request:
#### 88d25d8b73a78ffa2dff8623221c9acc44fec330
<pre>
[Liquid Glass] [macOS] Tab snapshots overlap with tab bar when swiping back/forwards
<a href="https://bugs.webkit.org/show_bug.cgi?id=295755">https://bugs.webkit.org/show_bug.cgi?id=295755</a>
<a href="https://rdar.apple.com/154032062">rdar://154032062</a>

Reviewed by Abrar Rahman Protyasha.

Tab snapshots are currently horizontally offset by the horizontal sidebar width, in the case where
the sidebar is open on macOS. This happens because the `m_swipeSnapshotLayer`&apos;s frame currently
overlaps with the left obscured inset area, as it&apos;s always placed at (0, 0) — bottom left on macOS,
relative to its parent `m_swipeLayer`. The y-coordinate happens to work out because Safari only
ever sets the top content inset in practice, so the bottom of the snapshot layer is always flush
against the bottom of the snapshot root layer.

However, when Liquid Glass is enabled, Safari now sets a left obscured content inset when the
sidebar is open. This requires the `m_swipeSnapshotLayer` to be positioned at `(sidebar width, 0)`,
and failure to do so results in the snapshot layer being tucked behind the sidebar.

To fix this, we set the `m_swipeSnapshotLayer`&apos;s frame to:
- `x` -&gt; `left obscured inset`
- `y` -&gt; `bottom obscured inset`
- `width` -&gt; `swipe area width, subtracting horizontal insets`
- `height` -&gt; `swipe area height, subtracting vertical insets`

Note: this applies to both cases where a custom swipe view is used (e.g. swiping between the start
page and a website), and the case where there&apos;s no custom swipe view (e.g. swiping between
websites).

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::itemForSwipeDirection const):

Add a helper method to return the relevant back/forward history list item when swiping in the given
direction.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::beginSwipeGesture):

Implement the main fix here; see above for more details.

(WebKit::ViewGestureController::beginSimulatedSwipeInDirectionForTesting):
(WebKit::ViewGestureController::completeSimulatedSwipeInDirectionForTesting):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::beginBackSwipeForTesting):

Add support for simulated navigation swipe testing hooks on macOS. We also adjust this codepath to
_ensure_ the gesture controller if it doesn&apos;t already exist, rather than bailing if the testing hook
is invoked while the controller hasn&apos;t been instantiated.

(WebKit::WebViewImpl::updateScrollPocket):

Drive-by fix: adjust the z-index to be the lowest `float` instead of the lowest `double`, to avoid a
benign warning in QuartzCore.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NavigationSwipeTests.mm: Renamed from Tools/TestWebKitAPI/Tests/ios/NavigationSwipeTests.mm.

Refactor this API test file:
1. Move it from `ios/` to `WebKitCocoa/`, so that we can put macOS-specific tests there as well.
2. Merge the single API test in `CustomSwipeViewTests` into this new suite.

(-[TestNavigationInteractiveTransition startInteractiveTransition:]):
(TestWebKitAPI::TEST(NavigationSwipeTests, RestoreFirstResponderAfterNavigationSwipe)):
(TestWebKitAPI::TEST(NavigationSwipeTests, DoNotBecomeFirstResponderAfterNavigationSwipeIfWebViewIsUnparented)):
(TestWebKitAPI::TEST(NavigationSwipeTests, DoNotAssertWhenSnapshottingZeroSizeView)):
(TestWebKitAPI::TEST(NavigationSwipeTests, WindowRelativeBoundsForCustomSwipeViews)):
(TestWebKitAPI::TEST(NavigationSwipeTests, SwipeSnapshotLayerBounds)):

Add a new API test that exercises the bug by directly checking the bounds of the swipe snapshot
layer after simulating a swipe backwards.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ObscuredContentInsets.mm:
(TestWebKitAPI::TEST(ObscuredContentInsets, TopOverhangColorExtensionLayer)):
* Tools/TestWebKitAPI/Tests/mac/CustomSwipeViewTests.mm: Removed.

Move this test into `NavigationSwipeTests`, and rename it appropriately.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(forEachCALayer):
(-[WKWebView forEachCALayer:]):
(-[WKWebView firstLayerWithName:]):

Make a couple adjustments to the testing helpers:
- Add a mechanism to stop iteration early when traversing the `CALayer` tree.
- Add a second helper method to grab a `CALayer` by name.

Canonical link: <a href="https://commits.webkit.org/297248@main">https://commits.webkit.org/297248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f5f526ab6510cb84ee6ed4c7f28047cbb0ed7e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21166 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117100 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113031 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84437 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114016 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25099 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99997 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64883 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24442 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60919 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94478 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18208 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119957 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38118 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28325 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93376 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38494 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93200 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23748 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38282 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16023 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38007 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43483 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37671 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41005 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39374 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1a3360daa38b9ca42fb3f431c50e76e03a647d83. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47867

aprotyas commented on Source/WebKit/UIProcess/ViewGestureController.h:
```diff
@@ -212,6 +212,9 @@ class ViewGestureController final : public IPC::MessageReceiver, public RefCount
     bool completeSimulatedSwipeInDirectionForTesting(SwipeDirection);
 
+protected:
+    RefPtr<WebBackForwardListItem> itemForSwipeDirection(SwipeDirection) const;
```
Can this be private?

aprotyas commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -7004,7 +7006,7 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
         captureView = [m_topScrollPocket captureView];
         [m_layerHostingView addSubview:captureView.get() positioned:NSWindowBelow relativeTo:nil];
-        [captureView layer].zPosition = -CGFLOAT_MAX;
+        [captureView layer].zPosition = -FLT_MAX;
```
I'm more of a `std::numeric_limits<float>::lowest()` person myself 😛

whsieh replied:
Works for me!

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/NavigationSwipeTests.mm:
```diff
@@ -95,3 +101,44 @@ - (void)startInteractiveTransition:(id <UIViewControllerContextTransitioning>)tr
+    [webView _setCustomSwipeViews:@[ customSwipeView.get() ]];
+    [webView _setCustomSwipeViewsObscuredContentInsets:NSEdgeInsetsMake(100, 200, 50, 50)];
+
+    __auto_type boundsForCustomSwipeView = [webView _windowRelativeBoundsForCustomSwipeViewsForTesting];
```
Nit: I know you're moving old  code around, but maybe just `auto`? They are identical in clang

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7f5f526ab6510cb84ee6ed4c7f28047cbb0ed7e3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47867

whsieh committed:
[Liquid Glass] [macOS] Tab snapshots overlap with tab bar when swiping back/forwards
https://bugs.webkit.org/show_bug.cgi?id=295755
rdar://154032062

Reviewed by Abrar Rahman Protyasha.

Tab snapshots are currently horizontally offset by the horizontal sidebar width, in the case where
the sidebar is open on macOS. This happens because the `m_swipeSnapshotLayer`'s frame currently
overlaps with the left obscured inset area, as it's always placed at (0, 0) — bottom left on macOS,
relative to its parent `m_swipeLayer`. The y-coordinate happens to work out because Safari only
ever sets the top content inset in practice, so the bottom of the snapshot layer is always flush
against the bottom of the snapshot root layer.

However, when Liquid Glass is enabled, Safari now sets a left obscured content inset when the
sidebar is open. This requires the `m_swipeSnapshotLayer` to be positioned at `(sidebar width, 0)`,
and failure to do so results in the snapshot layer being tucked behind the sidebar.

To fix this, we set the `m_swipeSnapshotLayer`'s frame to:
- `x` -> `left obscured inset`
- `y` -> `bottom obscured inset`
- `width` -> `swipe area width, subtracting horizontal insets`
- `height` -> `swipe area height, subtracting vertical insets`

Note: this applies to both cases where a custom swipe view is used (e.g. swiping between the start
page and a website), and the case where there's no custom swipe view (e.g. swiping between
websites).

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::itemForSwipeDirection const):

Add a helper method to return the relevant back/forward history list item when swiping in the given
direction.

* Source/WebKit/UIProcess/ViewGestureController.h:
* Source/WebKit/UIProcess/ios/ViewGestureControllerIOS.mm:
(WebKit::ViewGestureController::beginSwipeGesture):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::beginSwipeGesture):

Implement the main fix here; see above for more details.

(WebKit::ViewGestureController::beginSimulatedSwipeInDirectionForTesting):
(WebKit::ViewGestureController::completeSimulatedSwipeInDirectionForTesting):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::beginBackSwipeForTesting):

Add support for simulated navigation swipe testing hooks on macOS. We also adjust this codepath to
_ensure_ the gesture controller if it doesn't already exist, rather than bailing if the testing hook
is invoked while the controller hasn't been instantiated.

(WebKit::WebViewImpl::updateScrollPocket):

Drive-by fix: adjust the z-index to be the lowest `float` instead of the lowest `double`, to avoid a
benign warning in QuartzCore.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NavigationSwipeTests.mm: Renamed from Tools/TestWebKitAPI/Tests/ios/NavigationSwipeTests.mm.

Refactor this API test file:
1. Move it from `ios/` to `WebKitCocoa/`, so that we can put macOS-specific tests there as well.
2. Merge the single API test in `CustomSwipeViewTests` into this new suite.

(-[TestNavigationInteractiveTransition startInteractiveTransition:]):
(TestWebKitAPI::TEST(NavigationSwipeTests, RestoreFirstResponderAfterNavigationSwipe)):
(TestWebKitAPI::TEST(NavigationSwipeTests, DoNotBecomeFirstResponderAfterNavigationSwipeIfWebViewIsUnparented)):
(TestWebKitAPI::TEST(NavigationSwipeTests, DoNotAssertWhenSnapshottingZeroSizeView)):
(TestWebKitAPI::TEST(NavigationSwipeTests, WindowRelativeBoundsForCustomSwipeViews)):
(TestWebKitAPI::TEST(NavigationSwipeTests, SwipeSnapshotLayerBounds)):

Add a new API test that exercises the bug by directly checking the bounds of the swipe snapshot
layer after simulating a swipe backwards.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ObscuredContentInsets.mm:
(TestWebKitAPI::TEST(ObscuredContentInsets, TopOverhangColorExtensionLayer)):
* Tools/TestWebKitAPI/Tests/mac/CustomSwipeViewTests.mm: Removed.

Move this test into `NavigationSwipeTests`, and rename it appropriately.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(forEachCALayer):
(-[WKWebView forEachCALayer:]):
(-[WKWebView firstLayerWithName:]):

Make a couple adjustments to the testing helpers:
- Add a mechanism to stop iteration early when traversing the `CALayer` tree.
- Add a second helper method to grab a `CALayer` by name.

Canonical link: https://commits.webkit.org/297248@main
